feat(query): add partial query solver engine

fubuloubu opened the pull request:
### What I did

This PR adds (provisional) support for "partial queries", e.g. breaking large user queries through the Query Management System into sub-queries according to the most efficient manner of provisioning them (based on installed plugins, plugin state, service access, environment, etc.)

Most of the time, when running some long-term query over a range of history, often certain providers are not enough to fully provision the entire query as some parts may be missing. Examples of this include disk cache with `cryo` (or the not-fully-functional `ape-cache` 1st party plugin), indexing services who might be a few minutes behind head w/ their API, or other such application-specific limitations (rate limits? free tier limits? etc.)

Thanks to this plugin, it is now possible to better support these types of queries efficiently, and also allow plugins to reduce overhead cost of conversion of their outputs, thanks to additions to the API to allow specifying different manners of providing raw Ape model access, or dataframes using `narwhals`'s dataframe-agnostic API

fixes: #702
fixes: #1079
fixes: #2013

### How I did it

First off, all built-in queries were modified using a new baseclass approach to add the concept of "query indexes" as an abstract API (and then implemented across relevant queries). Then the concept of a "cursor" was added to represent a provider-specific subset of the query **without** evaluating it. The cursors aided in the development of a new algorithm that finds the most efficient subset of all query engine's cursor objects that "cover" the original query. Lastly, this new algorithm and all machinery to make best use of it was implemented into query subsystem API objects as well as the query manager, gated behind an environment variable `APE_ENABLE_EXPERIMENTAL_QUERY_BACKEND`

### How to verify it

**NOTE**: To use the new solver, first enable it using `APE_ENABLE_EXPERIMENTAL_QUERY_BACKEND=1`

All regular functionality of the framework as well as any query-capable plugins should continue to function properly. Support for new query functionality needs to be added prior to v0.9 where the changes will be made breaking.

Also note that while it should function without **any** breaking changes to current v0.8, it is intended to replace the functionality of the query system when v0.9 rolls around, dropping support for the using the old method of querying from 2nd/3rd party plugins. The drop of support should be transparent to them however, as the old API methods will no longer be called, and the new ones have defaults which should not hinder proper operation.

### Checklist

<!-- All PRs must complete the following checklist before being merged -->

- [x] All changes are completed
- [x] Change is covered in tests
- [x] Documentation is complete

fubuloubu committed:
refactor(query)!: add `.num_transactions` to `BlockTransactionQuery`

fubuloubu committed:
test: refactor old test that did not make sense

fubuloubu committed:
feat(query): add partial query support to Query API and `QueryManager`

- update all `QueryType`s and add `BaseCursorAPI` objects to
  `ape.api.query`
- add support to `QueryManager` for using the new partial query
  solver system
- add `APE_ENABLE_EXPERIMENTAL_QUERY_BACKEND` environment variable to
  enable the new (experimental) partial query solver system
- build support for default RPC queries through new partial query system

NOTE: Contains notes for v0.9 breaking changes *not made yet*

fubuloubu committed:
fix: Self is not available <3.11

fubuloubu commented:
maintainer's note: the idea behind this PR is to begin adding support for the new experimental query API to downstream 2nd/3rd party plugins in anticipation of v0.9, where legacy support will be dropped.

fubuloubu committed:
fix: remove Pandas as core dependency, add Narwhals (needed some refactor)

fubuloubu committed:
fix: used Py3.10+ Union type

fubuloubu committed:
feat(config): add ability to configure default query backend type

fubuloubu committed:
fix(query): forgot to set backend class because of temporary API issues

fubuloubu committed:
fix: add backport for pairwise when <3.10

fubuloubu committed:
fix: still using py3.10 union type

fubuloubu committed:
refactor: fix type heirarchy and other small bugs

antazoey reviewed (commented):
nice, consider my feedback then ill approve
also i would like to make sure this doesnt break ape-titanoboa

fubuloubu commented:
> nice, consider my feedback then ill approve
> also i would like to make sure this doesnt break ape-titanoboa

In general, I want to check all the other plugins by having a PR queued up before merging this.

fubuloubu commented on src/ape/api/config.py:
```diff
@@ -199,6 +200,17 @@ class DeploymentConfig(PluginConfig):
+class QueryConfig(PluginConfig):
+    """Add 'query:' key to your config."""
+
+    backend: DataframeImplementation = DataframeImplementation.PANDAS
```
So, narwhals is an abstraction package that has code to translate operations on your choice of dataframe implementation, so if you don't have pandas installed but pick pandas then it will raise (hence the config option)

Many people prefer polars over pandas, and this would allow us to support both without having to ship with either package (meaning you would install your preferred package and set the config to use it as default)

antazoey replied:
on option, we could have a dynamic default that prefers polars if it is installed but uses pandas otherwise

fubuloubu replied:
Yeah we can do that, because it's unlikely that someone would have 2 or more installed, and if they do they can just set it via config (or kwarg)

fubuloubu commented on src/ape/api/providers.py:
```diff
@@ -151,12 +151,21 @@ def transactions(self) -> list[TransactionAPI]:
         if self.hash is None:
-            # Unable to query transactions.
+            # NOTE: Only "unsealed" blocks do not have a hash
+            raise ProviderError("Unable to find block transactions: not sealed yet")
```
Not sure if there's a use case to work with that, but potentially thatmigjt be better

fubuloubu commented on tests/functional/test_block.py:
```diff
@@ -128,14 +128,3 @@ def test_model_validate_web3_block():
-    expected: list = []
-    assert actual == expected
-
-    # Ensure still works when hash is None (was a bug where this crashed).
```
I was trying to figure out why this test existed, thank you for commenting

antazoey commented on src/ape/api/providers.py:
```diff
@@ -151,12 +151,21 @@ def transactions(self) -> list[TransactionAPI]:
+
         if self.hash is None:
-            # Unable to query transactions.
+            # NOTE: Only "unsealed" blocks do not have a hash
```
I think we already made this change, but it was during development that led to the discovery of the bug leading to the test.

I think I am good with adjusting the test to catch an exception (instead of just completely deleting the test), as it was found with purpose.

fubuloubu committed:
fix: typing and formatting updates

fubuloubu committed:
refactor(QueryAPI): use new `QueryEngineAPI` class

antazoey reviewed (dismissed):
there's some unaddressed comments still but not going to worry about it

fubuloubu committed:
refactor(API): modify CursorAPI/QueryEngineAPI abstract methods

- rename BaseCursorAPI to CursorAPI
- remove singledispatchmethod from QueryEngineAPI.exec
- fill in some abstract methods on CursorAPI for easier impl
- updated DefaultQueryProvider

fubuloubu committed:
refactor(Ethereum): update trace query engine

fubuloubu committed:
refactor(Node): update Otterscan query plugin

fubuloubu committed:
WIP
